docker_cli/_utils.py

```diff
@@ -0,0 +1,256 @@
+    try:
+        base: Image = Image.build(  # type: ignore
+            tag=tag,
+            dockerfile_string=f"FROM {image_name}\nRUN mkdir {tag}",
+            stdout=stdout,
+            stderr=stderr,
+        )
+    except CalledProcessError:
+        raise ValueError(f"Image not found: {image_name}")
+
+    package_mgr = _package_manager_check(image=base)
+
+    if configure:
+        init_lines: str = (
+            "RUN " + str(package_mgr.generate_configure_command(stringify=True)) + "\n"
+        )
+    else:
+        init_lines = ""
+
+    url_program, url_init = _url_reader_check(image=base, package_mgr=package_mgr)
+    init_lines += url_init
+
+    run(split(f"docker image rm {tag}"), stdout=DEVNULL, stderr=DEVNULL)
+
+    return package_mgr, url_program, init_lines
+
+
+def _parse_cuda_info(cuda_version: str) -> Tuple[int, int]:
+    """
+    Turns a cuda version string into a major and minor version.
+
+    Parameters
+    ----------
+    cuda_version : str
+        The version string.
+
+    Returns
+    -------
+    Tuple[int, int]
+        The major and minor versions, in that order
+
+    Raises
+    ------
+    ValueError
+        If the input string does not encode a valid CUDA version number.
+    """
+    cuda_ver_pattern = re.compile(r"^(?P<major>[0-9]+)\." r"(?P<minor>[0-9]+)$")
+    cuda_ver_match = re.match(cuda_ver_pattern, cuda_version)
+    if not cuda_ver_match:
+        raise ValueError(f"Malformed CUDA version: {cuda_version}")
+    cuda_ver_match_groups = cuda_ver_match.groupdict()
+
+    cuda_major: int = int(cuda_ver_match_groups["major"])
+    cuda_minor: int = int(cuda_ver_match_groups["minor"])
+
+    return (cuda_major, cuda_minor)
+
+
+def _is_conda_pkg_name(line: str) -> bool:
+    """
+    Returns if a line appears to be an anaconda package URL
+
+    Used for filtering lockfiles
+
+    Parameters
+    ----------
+    line : str
+        The line
+
+    Returns
+    -------
+    bool
+        True if the line appears to be an anaconda package URL, false otherwise
+    """
+    conda_pkg_pattern = re.compile(r"^https:\/\/conda.anaconda.org\/\S*$")
+    return re.match(conda_pkg_pattern, line) is not None
+
+
+def test_image(image: Image, expression: str) -> bool:
+    """
+    Runs a test expression on an image.
+
+    Parameters
+    ----------
+    image : Image
+        The image.
+    expression : str
+        The expression.
+
+    Returns
+    -------
+    bool
+        True if "test" returned with status 0, False otherwise.
+    """
+    try:
+        image.run(f"test -d {expression}")
+        return True
+    except CalledProcessError as err:
+        if err.returncode == 1:
+            return False
+        else:
+            raise
+
+
+class UniqueGenerator:
+    """Generates random, threadsafe strings of lowercase letters and digits."""
+
+    lock = Lock()
+
+    @staticmethod
+    def generate(k: int = 10) -> str:
```

Comment:
I think the name "generate" is too vague -- let's give this a more self-describing name

```suggestion
    def generate_random_string(k: int = 10) -> str:
```